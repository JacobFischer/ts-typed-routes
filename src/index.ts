export * from "./parameter";
export * from "./route";
