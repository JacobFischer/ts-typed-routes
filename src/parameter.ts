/* eslint-disable import/export */// TS allows us to export multiple times for overloaded function signatures
/* eslint-disable @typescript-eslint/no-explicit-any */// any required here for Parameter types

/** A parameter in a typesafe route, used to name and convert to and from strings. */
export interface TypesafeRouteParameter<
    TName extends string,
    TType extends any,
> {
    name: TName;
    parser: (serialized: string) => TType;
    stringify: (value: TType) => string;
}

/**
 * A tuple for a route with a type parser.
 *
 * @param name - The name of the parameter in the route.
 * @param parser - The parser that takes a string and returns a new type. The return type of this function is
 * considered the typing of this parameter for future input types.
 * @param stringify - The function that parses the value from it's type to a serialized string.
 * @returns A tuple [name, parser].
 */
export function parameter<
    TType extends any,
    TName extends string,
>(
    name: TName,
    parser: (serialized: string) => TType,
    stringify?: (value: TType) => string,
): TypesafeRouteParameter<TName, TType>;

/**
 * A simple parameter in a route of type string.
 *
 * @param name - The name of the parameter in the route.
 * @returns A tuple of just [name].
 */
export function parameter<
    TName extends string,
>(
    name: TName,
): TypesafeRouteParameter<TName, string>;

/**
 * Creates a tuple for TypeScript to pickup the parameters in a route.
 *
 * @param name - The name of the parameter in the route.
 * @param parser - An optional type parser. Must be convertable to and from strings in the Route.
 * @param stringify - The function that parses the value from it's type to a serialized string.
 * @returns A tuple with at least the name of the parameter, and an optional second argument of the parser.
 */
export function parameter<
    TName extends string,
    TType extends any = string,
>(
    name: TName,
    parser?: (serialized: string) => TType,
    stringify?: (value: TType) => string,
): TypesafeRouteParameter<TName, TType> {
    return {
        name,
        parser: parser || (String as any), // TType defaults to string, so this is safe
        stringify: stringify || (String as any),
    };
}
