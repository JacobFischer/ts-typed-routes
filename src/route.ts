import type { BaseParameter } from './parameter';

type ValuesOf<T extends readonly unknown[]> = T[number];

/** Any parameter type withing a route segment. */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ParameterType = BaseParameter<string, any, boolean>;

/** A segment within a route. */
export type RouteSegment = string | ParameterType;

/** An immutable array that represents a route (path). */
export type RouteSegments = ReadonlyArray<RouteSegment>;

/** Extracts all the parameters parts of a route segment. */
export type AllRouteParameters<T extends RouteSegments> = Exclude<
  ValuesOf<T>,
  string
>;

/**
 * Extracts all the parameters of a route segment by if they are optional or
 * not.
 */
export type OnlyRouteParameters<
  T extends RouteSegments,
  TOptional extends boolean
> = Exclude<
  Exclude<ValuesOf<T>, string>,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  BaseParameter<string, any, TOptional>
>;

/** The shape of a path formatter function. */
export type PathFormatter<TSegments extends RouteSegments> = (
  // {} & string allows any string type, but preserves the named parameter
  // segments for IDEs to show developers
  // eslint-disable-next-line @typescript-eslint/ban-types
  parameterName: AllRouteParameters<TSegments>['name'] | ({} & string),
  optional: boolean,
) => string;

/**
 * Given a route segments array this will build the expected object shape with
 * optional parameters marked as such.
 */
export type ParametersObject<T extends RouteSegments> = {
  [key in OnlyRouteParameters<T, true>['name']]: ReturnType<
    Extract<OnlyRouteParameters<T, true>, { name: key }>['parser']
  >;
} &
  {
    [key in OnlyRouteParameters<T, false>['name']]?: ReturnType<
      Extract<OnlyRouteParameters<T, false>, { name: key }>['parser']
    >;
  };

/**
 * A immutable container of functions to format and parse segments of a route.
 */
export type Route<TSegments extends RouteSegments = RouteSegments> = {
  /**
   * Creates a **new** route by extending new string(s) and parameter(s) onto
   * the end of this route.
   *
   * @param segments - The new segments to concat to the end of this route.
   * @returns A new route that is this current route, with new segments on
   * the end. This route is not mutated.
   */
  readonly extend: <TSegments2 extends RouteSegment[]>(
    ...segments: TSegments2
  ) => Route<[...TSegments, ...TSegments2]>;

  /**
   * Formats the route with the named parameters and returns the result.
   *
   * @param parameters - The named parameters to use to format with.
   * @param options - Optional object of options.
   * @returns A string of the entire path with the values filled in.
   */
  readonly format: (
    parameters: Readonly<ParametersObject<TSegments>>,
    options?: {
      /** Optional character used to join all segments in the path. */
      joiner?: string;
      /**
       * Optional encoder to use _after_ parameters are formatted within the
       * route. Defaults to `encodeURIComponent`.
       */
      encoder?: (stringified: string) => string;
    },
  ) => string;

  /**
   * The default value(s) for parameters in this route.
   * This exists mostly as a way to check for key and types at runtime.
   */
  readonly defaults: Readonly<ParametersObject<TSegments>>;

  /** The array of segments that forms this route. */
  readonly segments: TSegments;

  /**.
   * Parses an object of key/value strings into their expected types
   *
   * @param obj - Key value pairs to parse.
   * @param options - Optional object of options to use when parsing.
   * @returns A new object of key values pairs, where each value was
   * de-serialized.
   */
  readonly parse: (
    obj: Record<string, string>,
    options?: {
      /** If this should use default values for keys not present. */
      useDefaults?: boolean;
      /**
       * Optional encoder to use _before_ parameters are passed to their parser
       * functions. Defaults to `decodeURIComponent`.
       */
      decoder?: (encoded: string) => string;
    },
  ) => ParametersObject<TSegments>;

  /**
   * Creates the raw path for this route, using parameter name in place.
   *
   * @param options - Options to use when formatting the path.
   * @param options.formatter - An optional formatter to invoke on each
   * parameter name. By default adds a ':' in front, and a '?' behind if
   * optional.
   * E.g. `optionalParameterName` -> `:optionalParameterName?`.
   * @returns The route's path with parameter names in place.
   */
  readonly path: (options?: {
    formatter?: PathFormatter<TSegments>;
    joiner?: string;
  }) => string;
};

/**
 * Creates a route of static strings and parameters.
 *
 * @param segments - The segments of the route, a mix of static strings and
 * parameters.
 * @returns A route object to help you build paths for this route.
 */
export function route<TSegments extends RouteSegment[]>(
  ...segments: TSegments
): Route<TSegments> {
  type Parameters = ParametersObject<TSegments>;
  const asString = (
    formatParameter: (p: ParameterType) => string | undefined,
    joiner = '/',
  ) =>
    segments
      .map((segment) =>
        typeof segment === 'string' ? segment : formatParameter(segment),
      )
      .filter((s) => s !== undefined)
      .join(joiner);

  const asObject = (
    reducer: (parameter: ParameterType) => Parameters[keyof Parameters],
  ) =>
    segments.reduce((parameters, segment) => {
      if (typeof segment !== 'string') {
        // it is a parameter segment, so get it's default value
        const key = segment.name as keyof Parameters;
        parameters[key] = reducer(segment);
      }
      return parameters;
    }, {} as Parameters);

  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  const defaults = asObject((parameter) => parameter.parser(''));

  const defaultFormatter: PathFormatter<TSegments> = (
    parameterName,
    optional,
  ) => `:${parameterName}${optional ? '?' : ''}`;

  return {
    extend(...newSegments) {
      return route(...segments, ...newSegments);
    },

    format(parameters, options) {
      const encoder = options?.encoder || encodeURIComponent;
      return asString(
        (p) =>
          p.optional && !(p.name in parameters)
            ? undefined
            : encoder(p.stringify(parameters[p.name as keyof Parameters])),
        options?.joiner,
      );
    },

    segments,
    defaults,

    parse(obj, options) {
      return asObject((parameter) => {
        const key = parameter.name as keyof Parameters;
        const value = obj[key];
        if (!(key in obj)) {
          if (options?.useDefaults || parameter.optional) {
            // eslint-disable-next-line @typescript-eslint/no-unsafe-return
            return defaults[key];
          }

          throw new Error(
            `route '${asString((p) =>
              defaultFormatter(p.name, p.optional),
            )}' cannot parse '${JSON.stringify(
              obj,
            )}', missing expected key '${String(key)}'`,
          );
        }

        const decoder = options?.decoder || decodeURIComponent;
        // eslint-disable-next-line @typescript-eslint/no-unsafe-return
        return parameter.parser(decoder(value));
      });
    },

    path(options) {
      const formatter = options?.formatter || defaultFormatter;
      return asString((p) => formatter(p.name, p.optional), options?.joiner);
    },
  };
}
