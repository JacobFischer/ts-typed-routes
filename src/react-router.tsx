/* NOTE: this part is VERY experimental */

try {
  require.resolve('react-router-dom');
} catch {
  throw new Error(`Cannot find module 'react-router-dom'.
You cannot use this part of 'ts-typed-routes' without these React packages installed too.`);
}

import * as React from 'react';
import {
  LinkProps,
  Link,
  NavLink,
  NavLinkProps,
  Redirect,
  RedirectProps,
  /*
  Route as ReactRouterRoute,
  RouteChildrenProps,
  RouteComponentProps,
  match,
  RouteProps,
  */
  useParams,
} from 'react-router-dom';
import type { BaseParameter } from './parameter';
import { route, Route } from './route';

// these are copy and pasted so they are not exported in the final bundle
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ParameterType = BaseParameter<string, any, boolean>;
type RouteSegment = string | ParameterType;

type RouteFormatOptions<T extends RouteSegment[]> = NonNullable<
  Parameters<Route<T>['format']>[1]
>;

type RouteParseOptions<T extends RouteSegment[]> = NonNullable<
  Parameters<Route<T>['parse']>[1]
>;

type FilterString<T extends unknown[]> = T extends []
  ? []
  : T extends [infer H, ...infer R]
  ? H extends string
    ? FilterString<R>
    : [H, ...FilterString<R>]
  : T;

type RouteParameters<T extends RouteSegment[]> = Route<T>['defaults'];
type ParametersProps<
  T extends RouteSegment[]
> = FilterString<T> extends never[]
  ? {
      parameters?: RouteParameters<T>;
    }
  : {
      parameters: RouteParameters<T>;
    };

/*
type MatchOverrides<T extends RouteSegment[]> = {
  match: Omit<match<Record<string, never>>, 'params'> & {
    params: RouteParameters<T>;
  };
};

type RouteComponentPropsOverrides<
  T extends RouteSegment[]
> = RouteComponentProps<Record<string, never>> & MatchOverrides<T>;

type RouteChildrenPropsOverrides<
  T extends RouteSegment[]
> = RouteChildrenProps<Record<string, never>> & MatchOverrides<T>;

type RoutePropsOverrides<T extends RouteSegment[]> = {
  component?: React.ComponentType<RouteComponentPropsOverrides<T>>;
  render?: (props: RouteComponentPropsOverrides<T>) => React.ReactNode;
  children?:
    | ((props: RouteChildrenPropsOverrides<T>) => React.ReactNode)
    | React.ReactNode;
};
*/

export interface ReactRoute<T extends RouteSegment[] = RouteSegment[]>
  extends Route<T> {
  readonly Link: React.ComponentType<
    Omit<LinkProps, 'to'> & ParametersProps<T> & RouteFormatOptions<T>
  >;

  readonly NavLink: React.ComponentType<
    Omit<NavLinkProps, 'to'> & ParametersProps<T> & RouteFormatOptions<T>
  >;

  readonly RedirectFrom: React.ComponentType<Omit<RedirectProps, 'from'>>;
  readonly RedirectTo: React.ComponentType<Omit<RedirectProps, 'to'>>;

  /*
  readonly Route: React.ComponentType<
    Omit<RouteProps, 'path' | 'render'> & RoutePropsOverrides<T>
  >;
  */

  readonly useParams: (options?: RouteParseOptions<T>) => RouteParameters<T>;

  /**
   * Creates a **new** ReactRoute by extending new string(s) and parameter(s)
   * onto the end of this route.
   *
   * @param segments - The new segments to concat to the end of this route.
   * @returns A new ReactRoute that is this current route, with new segments on
   * the end. This route is not mutated.
   */
  readonly extend: <TSegments2 extends RouteSegment[]>(
    ...segments: TSegments2
  ) => Route<[...T, ...TSegments2]>;
}

/**.
 * Creates a route of static strings and parameters.
 * This includes additional react-router functionality.
 *
 * @param segments - The segments of the route, a mix of static strings and
 * parameters.
 * @returns A reactRoute object to help you build paths for this route.
 */
export function reactRoute<T extends RouteSegment[]>(
  ...segments: T
): ReactRoute<T> {
  const baseRoute = route(...segments);
  return {
    ...baseRoute,
    Link: ({
      parameters = {} as RouteParameters<T>,
      joiner,
      encoder,
      ...props
    }) => (
      <Link
        to={baseRoute.format(parameters, { encoder, joiner })}
        {...props}
      />
    ),

    NavLink: ({
      parameters = {} as RouteParameters<T>,
      joiner,
      encoder,
      ...props
    }) => (
      <NavLink
        to={baseRoute.format(parameters, { encoder, joiner })}
        {...props}
      />
    ),

    RedirectFrom: (props) => <Redirect from={baseRoute.path()} {...props} />,

    RedirectTo: (props) => <Redirect to={baseRoute.path()} {...props} />,

    /*
    Route: ({ component, render, ...props }) => {
      const renderWrapped = render && ((...args) => render(...args));

      return <ReactRouterRoute path={baseRoute.path()} {...props} />;
    },
    */

    useParams: (options) => baseRoute.parse(useParams(), options),

    extend(...newSegments) {
      return reactRoute(...segments, ...newSegments);
    },
  };
}

export * from './parameter'; // re-export
