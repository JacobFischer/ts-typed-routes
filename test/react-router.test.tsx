import * as React from 'react';
import { Route, StaticRouter, StaticRouterContext } from 'react-router';
import { create } from 'react-test-renderer';
import { optionalParameter, parameter, reactRoute } from '../src/react-router';
import * as reactRouterExports from '../src/react-router';
import * as parameterExports from '../src/parameter';

describe('react-router', () => {
  it("should re-export parameter's exports", () => {
    const reactRouterExportsObj: Record<string, unknown> = reactRouterExports;
    for (const [key, value] of Object.entries(parameterExports)) {
      expect(key in reactRouterExportsObj).toBe(true);
      expect(reactRouterExportsObj[key]).toBe(value);
    }
  });

  describe('useParams hook', () => {
    it('should exist', () => {
      const simpleRoute = reactRoute('cars');
      expect(typeof simpleRoute.useParams).toBe('function');
    });

    it('should parse parameters via the hook', () => {
      const parameterRoute = reactRoute(
        '',
        parameter('str'),
        parameter('num', Number),
        optionalParameter('bool', Boolean),
      );

      const testValues: typeof parameterRoute.defaults = {
        str: 'some-string',
        num: 1337,
        bool: true,
      };

      const jestComponent = jest.fn(() => {
        const values = parameterRoute.useParams();

        expect(values.str).toStrictEqual(testValues.str);
        expect(values.num).toStrictEqual(testValues.num);
        expect(values.bool).toStrictEqual(testValues.bool);
        expect(values).toMatchObject(testValues);
        expect(values === testValues).toBe(false);

        return <>{JSON.stringify(values)}</>;
      });

      const Component: React.ComponentType = jestComponent;
      const rendered = create(
        <StaticRouter location={parameterRoute.format(testValues)}>
          <Route path={parameterRoute.path()}>
            <Component />
          </Route>
        </StaticRouter>,
      );

      expect(rendered.toJSON()).toMatchSnapshot();
      expect(jestComponent).toBeCalled();
    });
  });

  (['Link', 'NavLink'] as const).forEach((keyLink) =>
    describe(`<${keyLink} />`, () => {
      it('should exist', () => {
        const simpleRoute = reactRoute('foo', 'bar');
        expect(typeof simpleRoute[keyLink]).toBe('function');
      });

      it('should render a link', () => {
        const simpleRoute = reactRoute('foo', 'bar');
        const Component = simpleRoute[keyLink];
        const rendered = create(
          <StaticRouter>
            <Component />
          </StaticRouter>,
        );
        expect(rendered.toJSON()).toMatchSnapshot();

        const anchors = rendered.root.findAllByType('a');
        expect(anchors.length).toBe(1);
        const [anchor] = anchors;
        expect(anchor).toBeTruthy();
        expect(anchor.props.href).toBe('/foo/bar');
      });

      it('should not require parameters for string only routes', () => {
        const parameterRoute = reactRoute('alpha', 'beta', 'gamma');

        const Component = parameterRoute[keyLink];
        const rendered = create(
          <StaticRouter>
            <Component />
          </StaticRouter>,
        );

        expect(rendered.toJSON()).toMatchSnapshot();

        const anchors = rendered.root.findAllByType('a');
        expect(anchors.length).toBe(1);
        const [anchor] = anchors;
        expect(anchor).toBeTruthy();
        expect(anchor.props.href).toBe('/alpha/beta/gamma');
      });

      it('should require parameters', () => {
        const parameterRoute = reactRoute(
          'blog',
          parameter('post', String),
          optionalParameter('page', Number),
        );

        const Component = parameterRoute[keyLink];
        const rendered = create(
          <StaticRouter>
            <Component parameters={{ post: 'cool-id', page: 1 }} />
          </StaticRouter>,
        );

        expect(rendered.toJSON()).toMatchSnapshot();

        const anchors = rendered.root.findAllByType('a');
        expect(anchors.length).toBe(1);
        const [anchor] = anchors;
        expect(anchor).toBeTruthy();
        expect(anchor.props.href).toBe('/blog/cool-id/1');
      });
    }),
  );

  describe('<RedirectFrom />', () => {
    it('should exist', () => {
      const route = reactRoute();
      expect(typeof route.RedirectFrom).toBe('function');
    });

    it('should redirect from a formatted route', () => {
      const route = reactRoute('api', parameter('version'));
      const redirectTo = 'new-api/v2';
      const context: StaticRouterContext = {};
      const rendered = create(
        <StaticRouter location="/api/v1" context={context}>
          <route.RedirectFrom parameters={{ version: 'v1' }} to={redirectTo} />
        </StaticRouter>,
      );

      expect(rendered.toJSON()).toMatchSnapshot();
      expect(context.url).toStrictEqual(redirectTo);
    });

    it('should not require parameters when they are optional', () => {
      const route = reactRoute('', 'foo');
      const redirectTo = '/bar';
      const context: StaticRouterContext = {};
      const rendered = create(
        <StaticRouter location="/foo" context={context}>
          <route.RedirectFrom to={redirectTo} />
        </StaticRouter>,
      );

      expect(rendered.toJSON()).toMatchSnapshot();
      expect(context.url).toStrictEqual(redirectTo);
    });
  });

  describe('<RedirectFromPath />', () => {
    it('should exist', () => {
      const route = reactRoute();
      expect(typeof route.RedirectFromPath).toBe('function');
    });

    it('should redirect from a path', () => {
      const route = reactRoute(
        'anything',
        parameter('on'),
        parameter('thisPath'),
      );
      const redirectTo = 'new-path';
      const context: StaticRouterContext = {};
      const rendered = create(
        <StaticRouter location="/anything/var1/var2" context={context}>
          <route.RedirectFromPath to={redirectTo} />
        </StaticRouter>,
      );

      expect(rendered.toJSON()).toMatchSnapshot();
      expect(context.url).toStrictEqual(redirectTo);
    });
  });

  describe('<RedirectTo />', () => {
    it('should exist', () => {
      const route = reactRoute();
      expect(typeof route.RedirectTo).toBe('function');
    });

    it('should redirect to a formatted route', () => {
      const route = reactRoute('pizza', parameter('topping'));
      const redirectFrom = '/pepperoni-pizza';
      const context: StaticRouterContext = {};
      const rendered = create(
        <StaticRouter location={redirectFrom} context={context}>
          <route.RedirectTo
            from={redirectFrom}
            parameters={{ topping: 'pepperoni' }}
          />
        </StaticRouter>,
      );

      expect(rendered.toJSON()).toMatchSnapshot();
      expect(context.url).toStrictEqual('pizza/pepperoni');
    });

    it('should not require parameters when optional', () => {
      const route = reactRoute('', 'test');
      const redirectFrom = '/some-other-path';
      const context: StaticRouterContext = {};
      const rendered = create(
        <StaticRouter location={redirectFrom} context={context}>
          <route.RedirectTo from={redirectFrom} />
        </StaticRouter>,
      );

      expect(rendered.toJSON()).toMatchSnapshot();
      expect(context.url).toStrictEqual('/test');
    });
  });

  describe('<RedirectToPath />', () => {
    it('should exist', () => {
      const route = reactRoute();
      expect(typeof route.RedirectToPath).toBe('function');
    });

    it('should redirect to a path', () => {
      const route = reactRoute('users', 'new');
      const redirectFrom = 'new-user';
      const context: StaticRouterContext = {};
      const rendered = create(
        <StaticRouter location={redirectFrom} context={context}>
          <route.RedirectToPath from={redirectFrom} />
        </StaticRouter>,
      );

      expect(rendered.toJSON()).toMatchSnapshot();
      expect(context.url).toStrictEqual(route.path());
    });
  });

  describe('<Route />', () => {
    it('should exist', () => {
      expect(typeof reactRoute().Route).toBe('function');
    });

    it('should be the rendered route', () => {
      const route = reactRoute(
        '',
        'car',
        parameter('brand'),
        parameter('make'),
      );
      const rendered = create(
        <StaticRouter location="/car/tesla/model-3">
          <Route path="/car/tesla/model-s" />
          <route.Route>
            <strong>Found me!</strong>
          </route.Route>
          <Route>
            <span>Catch call route!</span>
          </Route>
        </StaticRouter>,
      );

      expect(rendered.toJSON()).toMatchSnapshot();

      const strongTags = rendered.root.findAllByType('strong');
      expect(strongTags.length).toBe(1);
      const [strongTag] = strongTags;
      expect(strongTag).toBeTruthy();
    });
  });

  describe('<RouteFormatted />', () => {
    it('should exist', () => {
      expect(typeof reactRoute().RouteFormatted).toBe('function');
    });

    it('should be the rendered route', () => {
      const route = reactRoute('/article', parameter('date'));
      const rendered = create(
        <StaticRouter location="/article/13-13-2033">
          <route.RouteFormatted parameters={{ date: '13-13-2033' }}>
            <em>Found me!</em>
          </route.RouteFormatted>
          <route.Route />
          <Route>
            <span>Catch call route!</span>
          </Route>
        </StaticRouter>,
      );

      expect(rendered.toJSON()).toMatchSnapshot();

      const emTags = rendered.root.findAllByType('em');
      expect(emTags.length).toBe(1);
      const [emTag] = emTags;
      expect(emTag).toBeTruthy();
    });

    it('should not require parameters when they are optional', () => {
      const route = reactRoute('', 'alpha');
      const rendered = create(
        <StaticRouter location="/alpha">
          <route.RouteFormatted>
            <em>Found me!</em>
          </route.RouteFormatted>
          <route.Route />
          <Route>
            <span>Catch call route!</span>
          </Route>
        </StaticRouter>,
      );

      expect(rendered.toJSON()).toMatchSnapshot();

      const emTags = rendered.root.findAllByType('em');
      expect(emTags.length).toBe(1);
      const [emTag] = emTags;
      expect(emTag).toBeTruthy();
    });
  });
});
