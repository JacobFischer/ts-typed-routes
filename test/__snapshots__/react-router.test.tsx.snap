// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-router <Link /> should render a link 1`] = `
<a
  href="/foo/bar"
  onClick={[Function]}
/>
`;

exports[`react-router <Link /> should require parameters 1`] = `
<a
  href="/blog/cool-id/1"
  onClick={[Function]}
/>
`;

exports[`react-router <NavLink /> should render a link 1`] = `
<a
  aria-current={null}
  href="/foo/bar"
  onClick={[Function]}
/>
`;

exports[`react-router <NavLink /> should require parameters 1`] = `
<a
  aria-current={null}
  href="/blog/cool-id/1"
  onClick={[Function]}
/>
`;

exports[`react-router <RedirectFrom /> should not require parameters when they are optional 1`] = `null`;

exports[`react-router <RedirectFrom /> should redirect from a formatted route 1`] = `null`;

exports[`react-router <RedirectFromPath /> should redirect from a path 1`] = `null`;

exports[`react-router <RedirectTo /> should not require parameters when optional 1`] = `null`;

exports[`react-router <RedirectTo /> should redirect to a formatted route 1`] = `null`;

exports[`react-router <RedirectToPath /> should redirect to a path 1`] = `null`;

exports[`react-router <Route /> should be the rendered route 1`] = `
Array [
  <strong>
    Found me!
  </strong>,
  <span>
    Catch call route!
  </span>,
]
`;

exports[`react-router <RouteFormatted /> should be the rendered route 1`] = `
Array [
  <em>
    Found me!
  </em>,
  <span>
    Catch call route!
  </span>,
]
`;

exports[`react-router <RouteFormatted /> should not require parameters when they are optional 1`] = `
Array [
  <em>
    Found me!
  </em>,
  <span>
    Catch call route!
  </span>,
]
`;

exports[`react-router useParams hook should parse parameters via the hook 1`] = `"{\\"str\\":\\"some-string\\",\\"num\\":1337,\\"bool\\":true}"`;
